### PR TITLE
add opensensemap layer

### LINKED ISSUE #58: Add layer: opensensemap.org
Just adding this quickly!!

### PR DESCRIPTION
**Pull request by rkpattnaik780**
A new layer has been added that gets data from opensensemap.org.

image

fixes #58

### CODE REVIEW

File: src/openSenseLayer.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      });
+    }
+  },
+
+  requestRegionData: function () {
+    var self = this;
+
+    (function () {
+      var $ = window.jQuery;
+      var url = "https://api.opensensemap.org/boxes";
```
Comment (jywarren) [inline src/openSenseLayer.js]: Hi! This looks great, just a small question -- 

Are we requesting only one region's worth of data at a time, and how are we submitting the lat/lon coordinates? Thanks!!

Also, this PR would make a great example for #136 -- thanks!

Comment (rkpattnaik780) [reply src/openSenseLayer.js]: Thank you @jywarren . I implemented the layer in this manner owing to the [api of opensensemap](https://api.opensensemap.org/). The current implementation has used two endpoints of the api :- 

- `/boxes` - Gets data of all the boxes.
- `/boxes/:boxId` - Gets data of a particular box.

Looking at the previously implemented layers, hitting the `/boxes` endpoint seemed sufficient to create the markers and their popups, but the problem was that it didn't return the values recorded by the sensors. So `/boxes/:boxId` had to be hit to get the values. 

So whenever an user opens the `Opensense layer`, the `/boxes` endpoint is first hit which provides the coordinates and the boxID, using which the markers are created. Whenever a marker is clicked, the popup is populated at that point only by making a get request to `boxes/:boxId` that returns the values recorded by the sensors for that particular region. That is why we can see a loading text before the data is loaded.

Please let me know if it requires changes. Thanks :)

**Comment by jywarren** [comment]
Thanks for the explanation -- sorry, just another question -- what defines
a box, and how do we generate a box id? Does it map to a lat/lon coordinate
region?

**Comment by rkpattnaik780** [comment]
@jywarren a `box` is short for a `sensebox` as defined by the `opensensemap API`. A `sensebox` is a group of sensors at a particular location. Each box object returned by `/boxes` comes with its coordinates and a unique `boxId`.  The current implementation stores this `boxId` to be used later for hitting the `/boxes/:boxId` route, which returns finer details of the box with that boxId. I would like to summarize my answers : -
1. A box is a collection of sensors at a particular location each having a list of sensors, a unique id and lat/lng coordinates.
2. We don't generate a boxId, we get it as a key of the box objects returned on hitting the `/boxes` route.
3. Yeah, each box maps to a lat/lng coordinate.

I hope this answers all the queries. Please let me know if anything is still unclear. Thanks :).

**Comment by jywarren** [comment]
OK, thanks! Does the code request all boxes, or are we restricting to only request boxes for the currently viewed region (and if so, could you point me at that section of code?) Thank you!

**Comment by rkpattnaik780** [comment]
@jywarren The code requests all the boxes when we select the layer. It had to be done cause there is no API endpoint that can fetch the boxes region-wise.
The code that gets all the boxes is in the file [dist/LeafletEnvironmentalLayers.js](https://github.com/publiclab/leaflet-environmental-layers/pull/142/files#diff-55fe3f5a2bd1a8fc3faafd6d407232bf) at line no. 26393. 
 I hope this answers your question, thanks :).

**Comment by jywarren** [comment]
Great, thank you so much!!!! 👍 We'd love some help with some of the organizational issues in #134 if you're interested!

**Comment by rkpattnaik780** [comment]
@jywarren thank you :). I will be working on it soon.

**Pull request merged**